### ISSUE TITLE
Package Name Map support as alternative to node module resolution

### ISSUE THREAD

**Issue by justinfagnani** [ry/deno#8]
Very interesting project! I love the web-compatibility goal.

One thing that would help there is support for bare module specifiers via package name maps: https://github.com/domenic/package-name-maps

This would allow for imports like:

```js
import * as _ from 'lodash';
```

Without any requirements on a specific package manager. It's our hope that npm, yarn and others will produce package-name maps during installation, giving you a bridge there. We have a reference implementation in progress here: https://github.com/domenic/package-name-maps/pull/37

**Comment by ry** [comment]
Interesting - I didn't know about that proposal. But in trying to seek a simpler module system, I would like to stick with very simple module specifiers for now (that is, exactly URLs).
On a related note: I'm struggling to decide if I should support extensionless URLs. It will make porting code much easier. On the other hand, it introduces the necessity to probe the system or remote server at multiple locations. Thoughts?

**Comment by qti3e** [comment]
@ry Actually I have an idea for that, we can use a meta-data in web pages like how `go get` does:
```html
<meta name="go-import" content="github.com/ry/deno git https://github.com/ry/deno.git">
```
something similar for deno would be:
```html
<meta name="deno-import" content="https://unpkg.com/deno_testing@0.0.5/testing.ts">
```

**Comment by TehShrike** [comment]
Extensionless URLs seem to have been a net negative for node.  Having to probe http/fs multiple times is a real expense if you need to optimize startup time, and it makes supporting new extensions later very difficult.

**Comment by qti3e** [comment]
@ry oh, my main point was html meta-tag on web-pages, not the git repo.
(in my example you can see the page maps to exactly one ts file, not a git repo.)
```html
<meta name="deno-import" content="https://unpkg.com/deno_testing@0.0.5/testing.ts">
```

**Comment by ry** [comment]
@qti3e I understood that already. That's a lot of work just to simplify linking to a file.

**Comment by justinfagnani** [comment]
@ry 

> On a related note: I'm struggling to decide if I should support extensionless URLs.

In the interest of web-compatibility and breaking local filesystem assumptions, I would absolutely not support this. A network fetcher just cannot do path searching like node can.

HTML currently throws if a specifiers is not a URL, or starts with `/`, `.`, or `../`. This reserves the "bare" specifier space for future proposals like package name maps, but the _existing_ valid specifiers are not going to be re-interpreted differently in the future. `./foo` won't change to resolve to `./foo.js`.

**Comment by justinfagnani** [comment]
> @TehShrike Yeah - I agree. It's unfortunate that a lot of code will have to be changed just to add an extension... but I think it's worth it - because browsers certainly are not going to do extension guessing - and it's just less code and more obvious.

I want to point out that there's a tools-based solution here. For Polymer tooling we support node-style module resolution in our dev server and build pipeline by just rewriting specifiers. It's also possible to do on package installation. A post-install step for deno could make incompatible packages work, much like the web has had to do for years now.

Then hopefully one day not all JavaScript will be written to node's specific peculiarities.

**Comment by ry** [comment]
@justinfagnani Yeah - in the off chance Deno moves beyond the prototype stage I would invest some time in making a source rewriter for porting code.
I'll close this - thanks for the helpful discussion.

**Comment by amio** [comment]
@ry 
> Deno just does http gets for source code

Seems that would be painful for installing a big dependencies graph, Deno have to download the full file to collect all packages(urls) it rely on, then go on one by one deep down the whole dep graph.

**Comment by ry** [comment]
@amio True, but you only have to do that once. It could very well be that it becomes very annoying.. but I'm betting it's just a small one-time cost.

**Comment by qti3e** [comment]
@amio Deno files are not large as you think they are...
node_modules is heavy because it downloads all the source code, not the bundle.
for your understanding, you can take a look at [this page](https://packagephobia.now.sh/result?p=propel@2.1.1) : )

(NPM will download 8.68MB but Deno will download only 121kB!)

**Comment by amio** [comment]
@qti3e Sure that's a huge win 🎉

While what I concern is, for a deep dependencies tree, eg:
```
a > b > c > d > e > f > g
```
Deno have to download these deps in series, rather than parallel. Roughly install time will depend on depth of the deps tree (assuming they are all same size, same network condition).

So the url-import-mechanism has pros & cons in installing speed, maybe overall it could still be faster than npm in practical. I think npm would be the baseline for "acceptable", and npm is not fast 😅

**Comment by qti3e** [comment]
@amio Don't worry!, implementing parallel package loading would be straightforward whenever we do "top-level await" : ) (I don't know if Ryan is going to accept that)

I will open another issue for that.

c.c: @ry

**Comment by reggi** [comment]
Regarding parallel package loading i think it should **not** be handled by `deno`.

Imagine there are n number of urls that are deeply nested `a > b > c > d > e > f > g`, deno would handle and install these dependencies and the first time it would be slow.

Ideally we can have a script that lists all of these dependencies as urls in a flat file:

```
https://unpkg.com/lodash@4.17.10/index.js
https://unpkg.com/bluebird@3.5.1/js/browser/bluebird.js
https://unpkg.com/shebang-regex@2.0.0/index.js
https://unpkg.com/fkill@5.3.0/index.js
```

This file could then be committed and a script can be used to download them all in parallel if they are not in the cache.

This would also ideally rectify a package that requires a local within itself

For instance if I used this:

```
import shebang from "https://unpkg.com/shebang-regex@2.0.0/index.js"
```

And this file looked like this:

```
import something from "./core.js"
```

Ideally this would have both files within the list:

```
https://unpkg.com/shebang-regex@2.0.0/index.js
https://unpkg.com/shebang-regex@2.0.0/core.js
```

**Comment by amio** [comment]
A deps listing file like `package.json`? 👍
Then it should handle versions like semver, then it would becomes `package.json` 🤔

**Comment by reggi** [comment]
It wouldn't be anything like package.json because it wouldn't be required.
These URLS are already canonical with the scope and version if they are
authored that way. This would simply be a faster way to download all
dependencies in parallel. No module resolution, no main, no bin, nothing
cute.

**Comment by amio** [comment]
I do agree with this:
> it should __not__ be handled by deno

Maybe it's more like `package-lock.json`
